text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+
+Many proposals for bitfields involved using macros to create structs, which
+would ultimately allow the underlying mechanisms for bitfield-based structs to
+be decided by the ecosystem, rather than the language itself.
+
+This is actually orthogonal to the proposal, as having language-level `uint<N>`
+types means that such macros would be able to return exactly the type necessary
+for each field. While something like `repr(bitpacked)` would be very far off in
+terms of stabilisation, these macros could simply get and set `uint<N>` values,
+and users could determine how to work with them without having to worry about
+checking for overflow or manually implementing code which has already been
+written.
+
+So, as far as this RFC is concerned, this is again orthogonal to the existing
+RFC.
+
+## Offering as a library
+
+Once const generics and specialisation are implemented and stable, almost all of
+this could be offered as a crate which offers `uint<N>` and `int<N>` types. I
+won't elaborate much on this because I feel that there are many other
+optimisations that can be added if this were a compiler feature, and that the
+standard library is simpler with them, but this is definitely worth adding here
+as an option.
+
+## Going without
+
+In the [Rust 2018 call for community blog posts], a fair number of Rustaceans
+expressed a desire for bitfields at the language level, for use in embedded
+code. So, while having bitfields is non-negotiable, this RFC's existence is
+definitely put into question.
+
+If this RFC were not accepted, then the most suitable alternative would be to
+develop a crate in the nursery which is suitable for creating bitflag-based
+structs in a way that offers robust error messages and reasonable overflow
+behaviour. However, this would not solve all of the other problems this RFC has
+mentioned, and this is likely not the ideal path.
+
+[Rust 2018 call for community blog posts]: https://readrust.net/rust-2018/
+[`bitfield` crate]: https://crates.io/crates/bitfield
+
+# Prior art
+[prior-art]: #prior-art
+
+At the time of writing, no known programming language offers this level of
+integer generalisation, and if this RFC were to be accepted, Rust would be the
+first.
+
+A few other ways of integer generalisation were explored in the alternatives
+section, like integer traits. A number of languages, notably Haskell and other
+functional languages, have generalised their integer types over a single trait.
+
+Additionally, Ada offers range typyes similar to those mentioned in the
+alternatives. I only noticed this near the completion of this RFC and haven't
+elaborated much on it.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+This RFC has a number of unresolved questions, many of which are probably suited
+to be solved during the development of this feature, rather than in this RFC.
+However, here are just a few:
+
+* Should `uN` and `iN` suffixes for integer literals, for arbitrary `N`, be
```

Comment:
An alternative to this would be to simply have `u<7>` and `i<42>` which would be almost as short...
Perhaps that's too short to be understandable? Chances are tho that given the fundamental nature of the types that people would remember it...

Reply:
Actually, I thought about this too. `uint` and `int` seem inconsistent with the other integer types somehow, so maybe `u` and `i` are the right choice?

Reply:
Also, either way, would we need to make a breaking change to make these identifiers reserved?

Reply:
I don't think we need to make breaking changes; you can always shadow the type with something else afaik.

Reply:
I'm adding this to the alternatives section but stating against it because `i` is such a common variable name.

Reply:
Well, it's a different namespace so there isn't a conflict.  The following ["works"](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=58ee3a271132832b7b22057689a8bd04):

```rust
struct i<i> { i: i };
let i: i<i32> = i { i: 4 };
```

(Without actually taking a position on whether `i` would be a good name for the type constructor.)